### PR TITLE
emit error when transform failed

### PR DESCRIPTION
**Pull request by franck34**
Sorry, i tried to make coverage better but i failed.

```
------------|----------|----------|----------|----------|----------------|
File        |  % Stmts | % Branch |  % Funcs |  % Lines |Uncovered Lines |
------------|----------|----------|----------|----------|----------------|
All files   |    99.22 |      100 |      100 |    99.22 |                |
 encoder.js |      100 |      100 |      100 |      100 |                |
 message.js |      100 |      100 |      100 |      100 |                |
 parser.js  |    96.88 |      100 |      100 |    96.88 |             82 |
------------|----------|----------|----------|----------|----------------|
```

My tests are failing too. Can you help ? thank you

### CODE REVIEW

File: test/tests/1-encoder.js [hunk 1]
```diff
@@ -70,6 +70,58 @@ describe( 'Encoder', () => {
...
+        }
+
+        translate() {
+          this.emit( 'error', new Error('myerror') );
+          return null;
+        }
+
+      }
+
+      const encoder  = new MessageEncoder({ objectMode: true });
```
Comment (kevincennis) [inline test/tests/1-encoder.js]:
```js
const encoder  = new MessageEncoder( {}, { objectMode: true } );
```

File: test/tests/1-encoder.js [hunk 2]
```diff
@@ -70,6 +70,58 @@ describe( 'Encoder', () => {
...
+
+      class MessageEncoder extends Encoder {
+
+        constructor() {
+          super();
+        }
+
+        translate() {
+          this.emit( 'error', new Error('myerror') );
+          return null;
```
Comment (kevincennis) [inline test/tests/1-encoder.js]:
```js
translate() {
  throw new Error('myerror');
}
```

File: test/tests/1-encoder.js [hunk 3]
```diff
@@ -70,6 +70,58 @@ describe( 'Encoder', () => {
...
+        }
```
Comment (kevincennis) [inline test/tests/1-encoder.js]:
```js
translate() {
  throw new Error('myerror');
}
```

File: test/tests/1-encoder.js [hunk 4]
```diff
@@ -70,6 +70,58 @@ describe( 'Encoder', () => {
...
+
+      }
+
+      const encoder  = new MessageEncoder({ objectMode: false });
```
Comment (kevincennis) [inline test/tests/1-encoder.js]:
```js
const encoder  = new MessageEncoder();
```

File: test/tests/2-parser.js [hunk 5]
```diff
@@ -133,6 +133,66 @@ describe( 'Parser', () => {
...
+      class MessageParser extends Parser {
+
+        constructor() {
+          super();
+        }
+
+        translate() {
+          this.emit( 'error', new Error('myerror') );
+          return null;
+        }
```
Comment (kevincennis) [inline test/tests/2-parser.js]:
```js
translate() {
  throw new Error('myerror');
}
```

File: test/tests/2-parser.js [hunk 6]
```diff
@@ -133,6 +133,66 @@ describe( 'Parser', () => {
...
+
+      }
+
+      const parser  = new MessageParser();
+
+      parser.on( 'error', err => {
+        console.log( err );
```
Comment (kevincennis) [inline test/tests/2-parser.js]: This should get removed.

File: test/tests/2-parser.js [hunk 7]
```diff
@@ -133,6 +133,66 @@ describe( 'Parser', () => {
...
+      class MessageParser extends Parser {
+
+        constructor() {
+          super();
+        }
+
+        translate() {
+          this.emit( 'error', new Error('myerror') );
+          return null;
+        }
```
Comment (kevincennis) [inline test/tests/2-parser.js]:
```js
translate() {
  throw new Error('myerror');
}
```

File: test/tests/2-parser.js [hunk 8]
```diff
@@ -133,6 +133,66 @@ describe( 'Parser', () => {
...
+
+      }
+
+      const parser  = new MessageParser();
+
+      parser.on( 'error', err => {
+        console.log( err );
```
Comment (kevincennis) [inline test/tests/2-parser.js]: This should be removed

File: test/tests/2-parser.js [hunk 9]
```diff
@@ -133,6 +133,66 @@ describe( 'Parser', () => {
...
+        done();
+      });
+
+      encoder.write('hello world');
+      parser.write( encoder.read() );
+    });
+
+    it( 'should emit `error` when error occured in transform using objectMode true', done => {
+      const Encoder = rewire( encoderPath );
+      const encoder = new Encoder({ objectMode: true });
```
Comment (kevincennis) [inline test/tests/2-parser.js]:
```js
const encoder = new Encoder( {}, { objectMode: true } );
```

File: test/tests/1-encoder.js (hunk 4 above)
Comment (kevincennis) [reply test/tests/1-encoder.js]: * `objectMode` is false by default

**Comment by kevincennis** [comment]
One other thing you might want to try that will make it clearer whether or not your tests are doing what you expect...

Set this up in each of your tests:

```js
const myError = new Error('myerror');
```

Then, inside your `translate` method:

```js
translate() {
  throw myError;
}
```

And finally, for your assertion

```js
encoder.on( 'error', err => {
  assert.equal( err, myError );
  done();
});
```

This way, you know that the `'error'` event that was emitted was the specific error you wanted. Otherwise, it could've been some other error you weren't expecting, which is what's happening with your tests now.

File: test/tests/1-encoder.js (hunk 1 above)
Comment (franck34) [reply test/tests/1-encoder.js]: ty

**Comment by franck34** [comment]
Thank you for all @kevincennis 

The Parser test with objectMode true fail now for another reason, and no idea on my side ...

```
 1) Parser #_transform should emit `error` when error occured in transform using objectMode true:
     RangeError [ERR_INDEX_OUT_OF_RANGE]: Index out of range
      at checkInt (buffer.js:1276:11)
      at Buffer.writeUInt32BE (buffer.js:1374:7)
      at Message.encode (lib/message.js:57:1230)
      at Encoder._transform (lib/encoder.js:25:206)
      at Encoder.Transform._read (_stream_transform.js:185:10)
      at Encoder.Transform._write (_stream_transform.js:173:12)
      at doWrite (_stream_writable.js:410:12)
      at writeOrBuffer (_stream_writable.js:396:5)
      at Encoder.Writable.write (_stream_writable.js:294:11)
      at Context.Parser (test/tests/2-parser.js:190:15)
```

**Comment by kevincennis** [comment]
Pretty sure the issue you're running into is that the `Encoder` class requires `translate` to return either a string or `Buffer`. Since the default `translate` method is an identity function and you're passing in an object, that assumption is broken. Basically, `Encoder` always needs to eventually produce a readable byte stream - but since it doesn't know how to turn arbitrary objects into buffers, you need to tell it how.

In this case, you can just subclass `Encoder` and give it a `translate` method that calls `JSON.stringify()` on its input, like this:

```js
it( 'should emit `error` when error occured in transform using objectMode true', done => {
  const Encoder = rewire( encoderPath );
  const Parser  = rewire( parserPath );
  const myError = new Error('myerror');

  class MessageParser extends Parser {

    constructor() {
      super();
    }

    translate() {
      throw myError;
    }

  }

  class MessageEncoder extends Encoder {

    translate( chunk ) {
      return JSON.stringify( chunk );
    }
  
  }

  const encoder = new Encoder({}, { objectMode: true });
  const parser  = new MessageParser();

  parser.on( 'error', err => {
    assert.equal( err, myError );
    done();
  });

  encoder.write({ myString: 'hello world' });
  parser.write( encoder.read() );
});
```

**Comment by franck34** [comment]
You win :D

**Comment by franck34** [comment]
Thank you kevin for your time. I learned a lot. Please notify when you publish a new npm version.

**Comment by kevincennis** [comment]
No, thank you.

A new version of fringe (`1.1.1`) has been published, as well as a new version of missive that pulls in this fix (`3.0.1`).

**Comment by franck34** [comment]
🥇 

FYI, I used Missive in https://github.com/sloki-project/sloki and https://github.com/sloki-project/sloki-node-client for the TCP binary protocol (a sort of jsonrpc, more light), at 18K ops/sec per clients. Ty again !

**Pull request merged**